### PR TITLE
CI: install virtualization operator to enable VM backup/restore testing

### PR DESCRIPTION
**Pull request by mrnold**
This pull request adds a place to put VM backup/restore testing, and adds a helper library for installing the Openshift Virtualization operator automatically. The only new test included is a check to see that the installation was successful, actual VM testing will be added in subsequent pull requests.

### CODE REVIEW

File: tests/e2e/scripts/virt_install.sh
Comment (mateusoliveira43) [inline tests/e2e/scripts/virt_install.sh:1]: the scripts in this folder are all bash, but this one seems a little more complex. Can it be written in go?

Comment (mrnold) [reply tests/e2e/scripts/virt_install.sh:1]: Yes, I have a Go version partially written. I can finish that for the final pull request.

**Comment by weshayutin** [comment]
```
Checking CNV CSV...
Error from server (NotFound): clusterserviceversions.operators.coreos.com "kubevirt-hyperconverged-operator.v4.13.6" not found
CSV phase is 
Failed installing CSV after two minutes!
```

**Comment by mrnold** [comment]
No more failures installing the virtualization operator, that's good. Hopefully the Mongo datamover failures are unrelated.

**Comment by mrnold** [comment]
I am only seeing consistent virtualization installation failures on 4.12, other failures look like unrelated tests that are also failing in other pull requests. I will look into the 4.12 failures, but otherwise this can be reviewed.

File: Makefile
Comment (mateusoliveira43) [inline Makefile:496]: These env vars are not needed anymore here, right?

Comment (mrnold) [reply Makefile]: Right, the *_settings.sh scripts don't need either of these.

File: tests/e2e/e2e_suite_test.go
Comment (mateusoliveira43) [inline tests/e2e/e2e_suite_test.go:41]: would put the same as in docs: Enable Virtual Machine backup/restore testing (default false)

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mateusoliveira43) [inline tests/e2e/virt_backup_restore_suite_test.go:13]: I would declare the variable and also move the os.Getenv logic to `tests/e2e/e2e_suite_test.go`

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mateusoliveira43) [inline tests/e2e/virt_backup_restore_suite_test.go:25]: should we also uninstall it?

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:47]: here I think you can use `kubeConf` from `tests/e2e/e2e_suite_test.go`

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:53]: would declare this client in `tests/e2e/e2e_suite_test.go`, for standardization

File: tests/e2e/lib/subscription_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/subscription_helpers.go:50]: are both streams necessary? are we interested in testing both?
Screenshot from 2023-12-28 14-20-35

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:153]: instead of hardcoding this, can we just test one?

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:268]: spec outside metadata? https://docs.openshift.com/container-platform/4.12/virt/install/installing-virt-cli.html#virt-deploying-operator-cli_installing-virt-cli

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:411]: 5 should be enough? 15 is too much

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [inline tests/e2e/lib/virt_helpers.go:184]: 4.12 jobs are failing here, maybe field does not exist there? https://github.com/kubevirt/hyperconverged-cluster-operator/commit/bab7b17a146b31ffbc4b749f6c70189245de64e8

I have 4.12 cluster, can try to debug (did not install kubevirt before)

**Review by mateusoliveira43 — CHANGES_REQUESTED** [review]
fix 4.12 installations

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:16]: I tried it this way first, but it got messy dealing with uninstalls stuck on finalizers and trying to only uninstall the operator if the test was the one to install it. It is possible, but I'm not sure it's worth the effort yet.

File: tests/e2e/lib/subscription_helpers.go
Comment (mrnold) [reply tests/e2e/lib/subscription_helpers.go]: I don't really know, I just saw that there was existing subscription-related code that did the same thing and copied it. I took this out to make it simpler for now.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:176]: Yes, that was the problem. I could probably create my own check that does the same thing, but it was easier just to skip these tests on versions older than 4.13.

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mateusoliveira43) [reply tests/e2e/virt_backup_restore_suite_test.go:16]: can we just add in the docs how to unistall it then? so someone running locally the test can return the cluster to original state after running the tests

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mateusoliveira43) [inline tests/e2e/virt_backup_restore_suite_test.go:21]: in `e2e_suite_test.go`we check if no error occurred when creating each client, so I think we do need to check if they are not nil here

File: tests/e2e/e2e_suite_test.go
Comment (mateusoliveira43) [inline tests/e2e/e2e_suite_test.go:102]: for standardization, add `ForSuiteRun` suffix to variable name

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [reply tests/e2e/lib/virt_helpers.go]: here I think `spec` is still inside `metadata`. It should be in the same level of apiVersion, Kind, and metadata, no?

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [reply tests/e2e/lib/virt_helpers.go:176]: just to we not forget, would create an issue to remove this code when we update CI and minimal version becomes 4.13

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go]: Was this "do not need to check"? I took these checks out, but I can add checks for every client instead if I read this the wrong way.

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:16]: Sure, I will see what I can do that is convenient for local tests.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:176]: Okay, if this code actually gets merged I will create an issue.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:196]: Same comment. I like simplicity of installOperatorGroup function, but possibly it should be just calling onother common method with name/namespace. Those functions are accessible within `&operatorsv1.*` but to simplify our code would be great to have such call in one place and the `installOperatorGroup` just a minimum required metadata to install such group. We may us then such call for other operatorgroups.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:219]: Similar to above comment.

Our CatalogSource, CatalogSourceNamespace could be another type, this way we can have easy way of installing operators from there? Just a thought, which may not be worth changing.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:246]: Could this schema be defined outside of the code or at least somewhere common together with other schema attributes which are not really functional.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:273]: I don't understand what it waits for? When reading the code I got it, just should be more straight forward from the desc.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:338]: Is that required? Why not to wait for csv when the csv is being created?

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:343]: I would use full  HyperConverged Operator in the desc as I had to go back to remember what the HCO stands for :)

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [inline tests/e2e/lib/virt_helpers.go:362]: Should this be part of v.checkHco? the v knows the namespace already and the checkHco() may simply return error if it's not installed within that NS. I don't think the ns check should be separate from checkHco() as one needs to be for other.

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mpryc) [inline tests/e2e/virt_backup_restore_suite_test.go:31]: The errors from many functions underneath are just as the Expect, should this be as well presented using log?

**Review by mpryc — COMMENTED** [review]
I was unable to test this on my cluster as I am still on 4.12.x and those tests are required to be ran on 4.13+. Some of the comments within the code itself.

```
VM backup and restore tests
/home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:12
  should verify virt installation
  /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:35
  > Enter [BeforeAll] VM backup and restore tests - /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:16 @ 01/12/24 11:41:58.07
2024/01/12 11:41:58 Getting packagemanifest...
2024/01/12 11:41:58 Extracting channels...
2024/01/12 11:41:58 Current CSV is: kubevirt-hyperconverged-operator.v4.12.8
2024/01/12 11:41:58 Current operator version is: 4.12.8
  [SKIPPED] Skipping virtualization testing on cluster version 4.12.8, minimum required is 4.13
  In [BeforeAll] at: /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:28 @ 01/12/24 11:41:58.6
  < Exit [BeforeAll] VM backup and restore tests - /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:16 @ 01/12/24 11:41:58.6 (530ms)
S [SKIPPED] [0.530 seconds]
VM backup and restore tests [BeforeAll]
/home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:16
  should verify virt installation
  /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/virt_backup_restore_suite_test.go:35
------------------------------
[AfterSuite] 
/home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:159
  > Enter [AfterSuite] TOP-LEVEL - /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:159 @ 01/12/24 11:41:58.6
2024/01/12 11:41:58 Deleting Velero CR
  [FAILED] Unexpected error:
      <*errors.errorString | 0xc00054f390>: 
      resource name may not be empty
      {
          s: "resource name may not be empty",
      }
  occurred
  In [AfterSuite] at: /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:168 @ 01/12/24 11:41:58.737
  < Exit [AfterSuite] TOP-LEVEL - /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:159 @ 01/12/24 11:41:58.737 (137ms)
[AfterSuite] [FAILED] [0.137 seconds]
[AfterSuite] 
/home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:159
------------------------------

Summarizing 1 Failure:
  [FAIL] [AfterSuite] 
  /home/migi/Development/Upstream/OADP/oadp-operator-virt-12.01.2024/tests/e2e/e2e_suite_test.go:168

Ran 0 of 32 Specs in 1.212 seconds
FAIL! -- 0 Passed | 0 Failed | 0 Pending | 32 Skipped
--- FAIL: TestOADPE2E (1.22s)
FAIL

Ginkgo ran 1 suite in 13.779622014s

Test Suite Failed

```

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mateusoliveira43) [reply tests/e2e/virt_backup_restore_suite_test.go]: yes yes, forget to add "not", thanks!

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [reply tests/e2e/lib/virt_helpers.go:186]: I would not do this kind of refactor in this PR. That can create questions as why not to use `tests/e2e/lib/kube_helpers.go` for this? It can be follow up PR

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:34]: In this file, I think GetVirtOperator, IsVirtInstalled, and EnsureVirtInstallation need to stay exported. They are called from virt_backup_restore_suite_test.go.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:123]: Where is this note? I don't see it in the current version: https://github.com/openshift/oadp-operator/blob/master/tests/e2e/lib/kube_helpers.go#L19

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:28]: I added a nil check for v.Version, if that's what you were asking for?

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:36]: I think this is just because it is a standalone test, not in a table of tests yet. I am sure there will be a table eventually, and this will be an Entry (or more likely get removed altogether).

File: tests/e2e/lib/subscription_helpers.go
Comment (mrnold) [reply tests/e2e/lib/subscription_helpers.go:56]: I think this only works because virt_helpers and subscription_helpers are in the same package? I am pretty sure Go doesn't otherwise let you add arbitrary functions to types like this, so it might not be applicable everywhere.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:128]: These ones (check*, install*, ensure*) are definitely not meant to be called from outside. GetOperatorSubscription is capitalized just to match the existing subscription helper code, but lower case does make more sense because it is only used privately.

The others that are still capitalized need to stay that way so they can be called from the test suite code.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:190]: When common_k8s_utils is merged, installNamespace would be a good candidate to put there. But probably named CreateNamespace or something. Maybe installOperatorGroup too, but probably not the others.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:200]: Yes, having this defer to a common CreateOperatorGroup would be nice.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:223]: This one I am less sure about, because it would likely look pretty much the same. It's pretty much just passing eight arguments in a Subscription to Create, and checking the result. It would be nice to have a whole factored-out Operator installation module, but I don't know that it's worth building that until we have more operators to install.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go]: I'm not totally sure what you are asking for here, but I moved the GVRs together to the top of the file.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go]: I added more words, and moved it to common_utils.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:320]: I don't know what is really required or not here. It's just the way I found to get the operator installed, directly translated from the command line version.

Comment (mpryc) [reply tests/e2e/lib/virt_helpers.go:320]: I am fine waiting here, just move it maybe inside the subscription function in it's body rather then separate function outside of the subscription. This way user will always wait for it to be ready. Otherwise consumer of function which creates subscription will need to remember to also wait for csv.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:344]: Right, the HCO check is the final thing that this uses to determine whether virtualization is good to go, and it depends on the namespace existing. But the public VirtOperator interface is only IsVirtInstalled and EnsureVirtInstallation, checkHco is an internal piece that is used correctly by both already.

The private functions are organized into check*, install*, and ensure*, where each * deals with just one thing. Then the public functions are responsible for using the private ones in the right order, which includes a namespace check (or creation) before doing anything with the HCO instance. So I don't see what would be the benefit of an extra namespace check in checkHco, unless we also want every other check function to also check all of its down dependencies.

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:32]: The virt_helpers stuff is reasonably well logged, and I was thinking ginkgo logs any error from this Expect nicer than I would anyway. Is there something else that needs to be logged?

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go:320]: I think that would be a good plan for a generic operator installer module, but here the only public interface is IsVirtInstalled and EnsureVirtInstallation. A consumer of these should not have to think about subscriptions, CSVs, or HCOs at all.

These could certainly be combined, but I am slightly more in favor of keeping it split into smaller chunks. If for whatever reason the Subscription creation failed, having them combined would mean you would have to wait for the longer timeout (5 minutes vs. 10 seconds) to fail to test. That's a small and unlikely failure mode, but enough of a reason to demand further convincing in favor of combining them.

File: tests/e2e/lib/virt_helpers.go
Comment (mateusoliveira43) [reply tests/e2e/lib/virt_helpers.go]: I think this function can be deleted, ginkgo has a built in function that does a very similar job

example on how to use it https://github.com/openshift/oadp-operator/blob/master/tests/e2e/lib/apps.go#L580

**Review by mateusoliveira43 — DISMISSED** [review]
Just need to check if WaitForCheck function can be replaced by eventually and add docs on how to uninstall hco operator

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mpryc) [reply tests/e2e/virt_backup_restore_suite_test.go:32]: That's fine. thanks for clarification.

File: tests/e2e/virt_backup_restore_suite_test.go
Comment (mpryc) [reply tests/e2e/virt_backup_restore_suite_test.go:28]: Thanks.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [reply tests/e2e/lib/virt_helpers.go:130]: You are right, don't know why I've checked so old version of this file:

https://github.com/openshift/oadp-operator/commit/a2d85f4f00d70db4d9a0555266713bd8785deb17#diff-06a8d2a8ef8552713ec00f3c9653d7c9727cf45a1b0bfa9c47a24929ba0e94e6L127-R78

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [reply tests/e2e/lib/virt_helpers.go:47]: ok.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [reply tests/e2e/lib/virt_helpers.go:344]: Thanks, makes sense.

File: tests/e2e/lib/virt_helpers.go
Comment (mpryc) [reply tests/e2e/lib/virt_helpers.go:320]: If we have long running concurrent operations then it makes sense to first ask for all of them and then check if we got them independently, so we save some blocking calls. But I am fine having a split here as well.

File: tests/e2e/lib/virt_helpers.go
Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go]: I didn't see ginkgo/gomega imported in any of the other *_helpers files, so I thought maybe it wasn't supposed to be used there for some reason. I did find (via nodeagent_helpers.go) a Kubernetes API machinery module called wait, with a PollImmediate function that does pretty much the same thing. I would rather use this one than the one from gomega, but either option is definitely better than writing a new one.

Comment (mrnold) [reply tests/e2e/lib/virt_helpers.go]: Okay, replaced this with PollImmediate from apimachinery.

**Review by mateusoliveira43 — APPROVED** [review]
just missing docs on how to uninstall hco operator

**Comment by mrnold** [comment]
> just missing docs on how to uninstall hco operator

I added some code to do the removal, just not automatically. Does this look any better than a doc? https://github.com/openshift/oadp-operator/pull/1257/commits/6db4abb185684b034785acb0aa33dc9864eed267

File: Makefile
Comment (mateusoliveira43) [inline Makefile:531]: I would undo this change. We used a similar approach with volsync in the past

https://github.com/openshift/oadp-operator/commit/b73b5c1207a1e199403b192c23ed49f76c1f450d#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L467-L470

https://github.com/openshift/oadp-operator/commit/a76c5dc2e2ef3706e9827933f45faddf0aae2c17#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L473-L493

but it did not install through test run, only through makefile. This way, we would have two options (test run and makefile). Would stick only with test run, removing this and `deploy/virt-operator.go` script. Since uninstall logic is already written, would add it to afterAll of virt test or add it in a follow up PR, after this one merges (PREFERRED)

Comment (mrnold) [reply Makefile]: Sure, I put an AfterAll to do this now instead, with a quick check to only remove it if it was the test that installed it in the first place.

File: deploy/virt-operator.go
Comment (mateusoliveira43) [inline deploy/virt-operator.go:1]: this file is still used?

**Review by mateusoliveira43 — DISMISSED** [review]

File: deploy/virt-operator.go
Comment (mrnold) [reply deploy/virt-operator.go:1]: Nope, I just forgot to remove it in the last commit.

**Comment by mpryc** [comment]
The `MySQL application two Vol CSI` found out flake, however 3 attempts of that test were not enough to run it.

The `Mongo application DATAMOVER` failure is not know flake, however it times out and more insight is needed to find out why.

**Review by kaovilai — APPROVED** [review]
Per PR description, the changes here LGTM.

**Pull request merged**
